File: scripts/legacy/export_config_from_db.py
#!/usr/bin/env python3
"""
DEPRECATED: Legacy export script for migrating configuration from the `key_values`
database table into a YAML file. The new configuration system is YAML-first and
the `key_values` table is deprecated/removed. This file is preserved for
historical migration use only and will not be actively maintained.
Usage: python3 scripts/legacy/export_config_from_db.py [path/to/filecodebox.db] [output.yaml]
"""
import sqlite3
import sys
import yaml
from pathlib import Path

DB_PATH = sys.argv[1] if len(sys.argv) > 1 else 'data/filecodebox.db'
OUT_PATH = sys.argv[2] if len(sys.argv) > 2 else 'config.generated.yaml'

conn = sqlite3.connect(DB_PATH)
cur = conn.cursor()
cur.execute("SELECT key, value FROM key_values")
rows = cur.fetchall()

cfg = {
    'base': {},
    'database': {},
    'storage': {},
    'user': {},
    'mcp': {},
    'ui': {},
}

for k, v in rows:
    if k == 'name': cfg['base']['name'] = v
    elif k == 'description': cfg['base']['description'] = v
    elif k == 'host': cfg['base']['host'] = v
    elif k == 'port':
        try:
            cfg['base']['port'] = int(v)
        except:
            cfg['base']['port'] = v
    elif k == 'data_path': cfg['base']['data_path'] = v
    elif k == 'production': cfg['base']['production'] = (v == 'true' or v == '1')
    elif k.startswith('database_'):
        sub = k.split('database_',1)[1]
        cfg['database'][sub] = int(v) if v.isdigit() else v
    elif k.startswith('storage.'):
        parts = k.split('.')
        if len(parts) >= 2:
            section = parts[1]
            if section not in cfg['storage']:
                cfg['storage'][section] = {}
            if len(parts) == 2:
                cfg['storage'][section] = v
            else:
                subkey = parts[2]
                cfg['storage'][section][subkey] = v
    elif k.startswith('user_'):
        sub = k.split('user_',1)[1]
        try:
            cfg['user'][sub] = int(v)
        except:
            cfg['user'][sub] = (v == 'true' or v == '1') if v in ['0','1','true','false'] else v
    elif k.startswith('mcp_'):
        sub = k.split('mcp_',1)[1]
        try:
            cfg['mcp'][sub] = int(v)
        except:
            cfg['mcp'][sub] = (v == 'true' or v == '1') if v in ['0','1','true','false'] else v
    elif k.startswith('theme') or k.startswith('notify') or k.startswith('page_'):
        cfg['ui'][k] = v
    else:
        cfg['ui'][k] = v

with open(OUT_PATH, 'w') as f:
    yaml.safe_dump(cfg, f, default_flow_style=False, sort_keys=False, allow_unicode=True)

print(f"Wrote {OUT_PATH} from {DB_PATH}")

File: scripts/legacy/export_config_from_db.go
package main

import (
	"database/sql"
	"flag"
	"fmt"
	"log"
	"os"

	_ "github.com/mattn/go-sqlite3"
	"gopkg.in/yaml.v3"
)

// DEPRECATED: Legacy migration tool to export key_values table to YAML.
func main() {
	dbPath := flag.String("db", "data/filecodebox.db", "path to sqlite db")
	out := flag.String("out", "config.generated.yaml", "output yaml file")
	flag.Parse()

	if _, err := os.Stat(*dbPath); err != nil {
		log.Fatalf("db not found: %v", err)
	}

	db, err := sql.Open("sqlite3", *dbPath)
	if err != nil {
		log.Fatalf("open db: %v", err)
	}
	defer db.Close()

	r, err := db.Query("SELECT key, value FROM key_values")
	if err != nil {
		log.Fatalf("query: %v", err)
	}
	defer r.Close()

	cfg := make(map[string]map[string]interface{})
	cfg["base"] = map[string]interface{}{}
	cfg["database"] = map[string]interface{}{}
	cfg["storage"] = map[string]interface{}{}
	cfg["user"] = map[string]interface{}{}
	cfg["mcp"] = map[string]interface{}{}
	cfg["ui"] = map[string]interface{}{}

	for r.Next() {
		var k, v string
		if err := r.Scan(&k, &v); err != nil {
			log.Printf("scan err: %v", err)
			continue
		}

		if k == "name" || k == "description" || k == "host" || k == "port" || k == "data_path" || k == "production" {
			cfg["base"][k] = parseValue(v)
			continue
		}
		if len(k) >= 9 && k[:9] == "database_" {
			sub := k[9:]
			cfg["database"][sub] = parseValue(v)
			continue
		}
		if len(k) >= 6 && k[:5] == "user_" {
			sub := k[5:]
			cfg["user"][sub] = parseValue(v)
			continue
		}
		if len(k) >= 4 && k[:4] == "mcp_" {
			sub := k[4:]
			cfg["mcp"][sub] = parseValue(v)
			continue
		}
		if len(k) >= 8 && k[:8] == "storage." {
			parts := splitN(k, '.', 3)
			if len(parts) >= 2 {
				section := parts[1]
				if len(parts) == 2 {
					cfg["storage"][section] = parseValue(v)
				} else {
					m, ok := cfg["storage"][section].(map[string]interface{})
					if !ok {
						m = map[string]interface{}{}
					}
					m[parts[2]] = parseValue(v)
					cfg["storage"][section] = m
				}
			}
			continue
		}
		cfg["ui"][k] = parseValue(v)
	}

	outF, err := os.Create(*out)
	if err != nil {
		log.Fatalf("create out: %v", err)
	}
	enc := yaml.NewEncoder(outF)
	enc.SetIndent(2)
	if err := enc.Encode(cfg); err != nil {
		log.Fatalf("encode yaml: %v", err)
	}
	outF.Close()
	fmt.Printf("wrote %s\n", *out)
}

func splitN(s string, sep byte, n int) []string {
	res := []string{}
	cur := ""
	count := 0
	for i := 0; i < len(s); i++ {
		if s[i] == sep {
			res = append(res, cur)
			cur = ""
			count++
			if count >= n-1 {
				res = append(res, s[i+1:])
				return res
			}
		} else {
			cur += string(s[i])
		}
	}
	res = append(res, cur)
	return res
}

func parseValue(v string) interface{} {
	var i int
	_, err := fmt.Sscanf(v, "%d", &i)
	if err == nil {
		return i
	}
	if v == "true" || v == "1" {
		return true
	}
	if v == "false" || v == "0" {
		return false
	}
	return v
}
